src/bosh-dns/dns/server/monitoring/metrics_server_test.go

```diff
@@ -0,0 +1,103 @@
+package monitoring_test
+
+import (
+	"bosh-dns/dns/server/internal/internalfakes"
+	"bosh-dns/dns/server/monitoring"
+	"bosh-dns/dns/server/monitoring/monitoringfakes"
+	"context"
+	"errors"
+	"net"
+
+	"github.com/cloudfoundry/bosh-utils/logger/loggerfakes"
+	"github.com/coredns/coredns/plugin/metrics"
+	"github.com/miekg/dns"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("MetricsServerWrapper", func() {
+	var (
+		shutdown                 chan struct{}
+		fakeLogger               *loggerfakes.FakeLogger
+		fakeMetricsServerWrapper *monitoringfakes.FakeCoreDNSMetricsServer
+	)
+
+	Describe("Run", func() {
+
+		BeforeEach(func() {
+			shutdown = make(chan struct{})
+			fakeMetricsServerWrapper = &monitoringfakes.FakeCoreDNSMetricsServer{}
+			fakeLogger = &loggerfakes.FakeLogger{}
+		})
+
+		Context("when no errors happen", func() {
+			BeforeEach(func() {
+				fakeMetricsServerWrapper.OnStartupStub = func() error {
+					close(shutdown)
+					return nil
+				}
+			})
+
+			It("starts and stops the server properly", func() {
+				err := monitoring.NewMetricsServerWrapper(fakeLogger, fakeMetricsServerWrapper).Run(shutdown)
+
+				Expect(err).ToNot(HaveOccurred())
+				Expect(fakeMetricsServerWrapper.OnFinalShutdownCallCount()).To(Equal(1))
+			})
+		})
+
+		Context("when start fails", func() {
+			BeforeEach(func() {
+				fakeMetricsServerWrapper.OnStartupStub = func() error {
+					close(shutdown)
+					return errors.New("")
+				}
+			})
+
+			It("returns error and doesn't wait for stop", func() {
+				err := monitoring.NewMetricsServerWrapper(fakeLogger, fakeMetricsServerWrapper).Run(shutdown)
+
+				Expect(err).To(HaveOccurred())
+				Expect(fakeMetricsServerWrapper.OnFinalShutdownCallCount()).To(Equal(0))
+			})
+		})
+	})
+
+	Describe("Report", func() {
+		var (
+			metritcsReporter monitoring.MetricsReporter
+			metricsServer    monitoring.CoreDNSMetricsServer
+			fakeWriter       *internalfakes.FakeResponseWriter
+			response         *dns.Msg
+		)
+
+		BeforeEach(func() {
+			fakeWriter = &internalfakes.FakeResponseWriter{}
+			metricsServer = monitoring.MetricsServer("127.0.0.1:53088")
+			metritcsReporter = monitoring.NewMetricsServerWrapper(fakeLogger, metricsServer).MetricsReporter()
+			response = &dns.Msg{
```

Comment:
This response doesn't appear to be used? Should there be a test asserting on a response metric when the fakeWriter is invoked?

Reply:
Good catch. I removed the unneeded response and added an expectation for the response count. There are no other response metrics at this point because we don't have a dns server serving the requests. I will keep the test with this setup because in it I'm really interested wether the handler is registered correctly.